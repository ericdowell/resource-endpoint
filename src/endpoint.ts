import { AxiosError, AxiosRequestConfig, AxiosResponse, ParamsSerializerOptions } from 'axios'
import { Options, globalOptions } from './options'
import qs, { IStringifyOptions } from 'qs'
import urljoin from 'url-join'

export class Endpoint {
  protected localOptions?: Options

  setLocalOptions(options: Options | undefined): this {
    this.localOptions = options
    return this
  }

  // Allowing global options to configure all Endpoint classes.
  // Optionally override options with this.localOptions via this.setLocalOptions
  get options(): Options {
    return this.localOptions ?? globalOptions
  }

  // Used in merging together AxiosRequestConfig values.
  // These are considered fallback values that
  // can be overridden by calling config.
  get config(): AxiosRequestConfig {
    return {}
  }

  // This contains the protocol and domain, aka location.origin.
  // e.g. https://example.com, http://localhost:3000, etc.
  get origin(): string {
    return this.options.origin
  }

  // This is the middle of the url path, usually a group prefix.
  // e.g. api/v1, user/settings, etc.
  get path(): string {
    return this.options.path
  }

  // This is the last part of the url path, usually the resource name.
  // e.g. "user" or "profile"
  get resource(): string {
    return ''
  }

  // Used to set baseURL property in AxiosRequestConfig.
  get baseURL(): string {
    return urljoin(this.origin, this.path, this.resource)
  }

  get stringifyOptions(): IStringifyOptions {
    return { arrayFormat: 'brackets' }
  }

  get paramsSerializer(): ParamsSerializerOptions {
    return {
      serialize: (params) => qs.stringify(params, this.stringifyOptions),
    }
  }

  // Returns AxiosRequestConfig passed to axios.request.
  requestConfig(requestConfig: AxiosRequestConfig): AxiosRequestConfig {
    const config: AxiosRequestConfig = { ...this.config, ...requestConfig }
    config.baseURL = config.baseURL ?? this.baseURL
    config.headers = { ...this.config.headers, ...config.headers }
    config.paramsSerializer = config.paramsSerializer ?? this.paramsSerializer
    return config
  }

  // General request method that is used by all HTTP calls.
  async request<T = any, R = AxiosResponse<T>>(requestConfig: AxiosRequestConfig): Promise<R> {
    try {
      return await this.options.axios.request<T, R>(this.requestConfig(requestConfig))
    } catch (error: any) {
      return this.handleError<T, R>(error)
    }
  }

  async handleError<T = any, R = AxiosResponse<T>>(error: AxiosError<T>): Promise<R | never> {
    throw error
  }
}
