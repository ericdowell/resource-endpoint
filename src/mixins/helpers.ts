export const defaultErrorMessage = 'An unexpected error has occurred. Please try again.'

export const errorsBlock = (message: string): { message: string } => ({ message })
