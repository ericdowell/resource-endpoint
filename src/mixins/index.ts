export * from './api'
export * from './auth'
export * from './axiosError'
export * from './form'
export * from './handleApi'
export * from './handleError'
export * from './sessionCsrfCookie'
export * from './types'
export * from './user'
