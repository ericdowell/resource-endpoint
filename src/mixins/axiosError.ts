import { Constructor } from './types'
import axios, { AxiosError } from 'axios'
import { safeResponseData } from '../helpers'
import { defaultErrorMessage as constDefaultErrorMessage, errorsBlock as baseErrorsBlock } from './helpers'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function AxiosErrorMixin<T extends Constructor<any>>(superClass: T) {
  return class extends superClass {
    get defaultErrorMessage(): string {
      return constDefaultErrorMessage
    }

    errorsBlock(message: string): { message: string } {
      return baseErrorsBlock(message)
    }

    handleError(error: AxiosError | Error): never {
      let message = this.defaultErrorMessage
      if (!axios.isAxiosError(error)) {
        const axiosErr = new AxiosError(error.message)
        axiosErr.response = {
          headers: {},
          config: {},
          status: 500,
          statusText: 'Internal Server Error',
          data: {
            message,
            errors: this.errorsBlock(message),
          },
        }
        throw axiosErr
      }
      const { errors, message: responseMessage } = safeResponseData<Record<string, unknown>>(error.response)
      // Matches what we want from our response.
      if (errors && typeof errors === 'object') {
        throw error
      } else if (typeof responseMessage === 'string') {
        message = responseMessage
      }
      error.response = {
        headers: error?.response?.headers ?? {},
        config: error?.config ?? error?.response?.config,
        status: error?.response?.status ?? 500,
        statusText: error?.response?.statusText ?? 'Internal Server Error',
        data: {
          message,
          errors: this.errorsBlock(message),
        },
      }
      throw error
    }
  }
}
