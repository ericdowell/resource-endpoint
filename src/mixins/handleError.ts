import { AxiosError, AxiosResponse } from 'axios'
import { Constructor } from './types'
import { defaultErrorMessage as constDefaultErrorMessage, errorsBlock as baseErrorsBlock } from './helpers'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function HandleErrorMixin<T extends Constructor<any>>(superClass: T) {
  return class extends superClass {
    /** @protected */
    shouldThrowError = false

    throwResponseError(): void {
      this.shouldThrowError = true
    }

    get defaultErrorMessage(): string {
      return constDefaultErrorMessage
    }

    errorsBlock(message: string): { message: string } {
      return baseErrorsBlock(message)
    }

    handleError<T = any>(error: AxiosError<T> | any): AxiosResponse<T | { message?: string; errors: any }> {
      if (this.shouldThrowError) {
        throw error
      }
      let message = this.defaultErrorMessage
      if (error?.response?.data?.errors) {
        return error.response
      } else if (typeof error?.response?.data?.message === 'string') {
        message = error.response.data.message
      }
      return {
        headers: error?.response?.headers,
        config: error?.config ?? error?.response?.config,
        status: error?.response?.status ?? 500,
        statusText: error?.response?.statusText ?? 'Internal Server Error',
        data: {
          message,
          errors: this.errorsBlock(message),
        },
      }
    }
  }
}
