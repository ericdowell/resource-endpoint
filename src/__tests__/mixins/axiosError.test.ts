import axios, { AxiosError, AxiosResponse } from 'axios'
import { AxiosErrorMixin } from '../../mixins'
import { Endpoint } from '../../endpoint'
import { BasicMock } from '../mock/axios'

jest.spyOn(axios, 'request').mockImplementation(BasicMock)
// eslint-disable-next-line @typescript-eslint/no-empty-function
jest.spyOn(global.console, 'log').mockImplementation(() => {})

class TestEndpoint extends AxiosErrorMixin(Endpoint) {}

const response = {
  config: {},
  status: 422,
  statusText: 'Unprocessable Entity',
  headers: {},
  data: {},
}

type TransformInput = (axiosResponse?: Partial<AxiosResponse>) => AxiosError

const noTransform = (input: any): any => input

const makeAxiosError: TransformInput = (axiosResponse?: Partial<AxiosResponse>): AxiosError => {
  const err = new AxiosError()
  if (response) {
    err.response = {
      data: undefined,
      status: 500,
      statusText: 'Testing',
      headers: {},
      config: {},
      ...axiosResponse,
    }
  }
  return err
}

describe(`${AxiosErrorMixin.name}`, (): void => {
  it.each([
    [undefined, noTransform],
    [new Error('foo bar issue'), noTransform],
    [{ data: undefined }, makeAxiosError],
    [{ data: { errors: { key: [] } } }, makeAxiosError],
    [{ ...response, data: { message: 'hello foo bar' } }, makeAxiosError],
    [{ ...response, status: 400, statusText: 'Bad Request' }, makeAxiosError],
  ])(
    'handleError throw AxiosError based on input %p',
    async (error: any, transform: TransformInput): Promise<void> => {
      expect.assertions(1)
      const expectThrow = async (): Promise<never> => new TestEndpoint().handleError(transform(error))
      await expect(expectThrow()).rejects.toMatchSnapshot()
    },
  )
})
