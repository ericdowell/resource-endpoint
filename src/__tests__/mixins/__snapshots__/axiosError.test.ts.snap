// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxiosErrorMixin handleError throw AxiosError based on input [Error: foo bar issue] 1`] = `[AxiosError: foo bar issue]`;

exports[`AxiosErrorMixin handleError throw AxiosError based on input {"config": [Object], "data": [Object], "headers": [Object], "status": 400, "statusText": "Bad Request"} 1`] = `[AxiosError: An unexpected error has occurred. Please try again.]`;

exports[`AxiosErrorMixin handleError throw AxiosError based on input {"config": [Object], "data": [Object], "headers": [Object], "status": 422, "statusText": "Unprocessable Entity"} 1`] = `[AxiosError: hello foo bar]`;

exports[`AxiosErrorMixin handleError throw AxiosError based on input {"data": [Object]} 1`] = `[AxiosError: An unexpected error has occurred. Please try again.]`;

exports[`AxiosErrorMixin handleError throw AxiosError based on input {"data": undefined} 1`] = `[AxiosError: An unexpected error has occurred. Please try again.]`;

exports[`AxiosErrorMixin handleError throw AxiosError based on input undefined 1`] = `[AxiosError]`;
